epoxy-paging/src/main/java/com/airbnb/epoxy/paging/CachingPagingEpoxyController.kt

```diff
@@ -0,0 +1,114 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.airbnb.epoxy.paging
+
+import android.arch.paging.PagedList
+import android.os.Handler
+import android.support.v7.util.DiffUtil
+import com.airbnb.epoxy.EpoxyController
+import com.airbnb.epoxy.EpoxyModel
+import com.airbnb.epoxy.EpoxyViewHolder
+
+/**
+ * An [EpoxyController] that can work with a [PagedList].
+ *
+ * Internally, it caches the model for each item in the [PagedList]. You should override
+ * [buildItemModel] method to build the model for the given item. Since [PagedList] might include
+ * `null` items if placeholders are enabled, this method needs to handle `null` values in the list.
+ *
+ * By default, the model for each item is added  to the model list. To change this behavior (to
+ * filter items or inject extra items), you can override [addModels] function and manually add built
+ * models.
+ */
+abstract class CachingPagingEpoxyController<T>(
+  /**
+   * The handler to use for build models.
+   */
+  modelBuildingHandler: Handler = EpoxyController.defaultModelBuildingHandler,
+  /**
+   * The handler to use when calculating the diff between built model lists
+   */
+  diffingHandler: Handler = EpoxyController.defaultDiffingHandler,
+  /**
+   * [CachingPagingEpoxyController] uses an [DiffUtil.ItemCallback] to detect changes between
+   * [PagedList]s. By default, it relies on simple object equality but you can provide a custom
+   * one if you don't use all fields in the object in your models.
+   */
+  itemDiffCallback: DiffUtil.ItemCallback<T> = DEFAULT_ITEM_DIFF_CALLBACK as DiffUtil.ItemCallback<T>
+) : EpoxyController(modelBuildingHandler, diffingHandler) {
+  // this is where we keep the already built models
+  private val modelCache = PagedListModelCache(
+    modelBuilder = { pos, item ->
+      buildItemModel(pos, item)
+    },
+    rebuildCallback = {
+      requestDelayedModelBuild(0)
+    },
+    itemDiffCallback = itemDiffCallback,
+    modelBuildingHandler = modelBuildingHandler
+  )
+
+  final override fun buildModels() {
+    addModels(modelCache.getModels())
+  }
+
+  /**
+   * This function adds all built models to the adapter. You can override this method to add extra
+   * items into the model list or remove some.
+   */
+  open fun addModels(models: List<EpoxyModel<*>>) {
+    super.add(models)
+  }
+
+  /**
+   * Builds the model for a given item. This must return a single model for each item. If you want
+   * to inject headers etc, you can override [addModels] function.
+   */
+  abstract fun buildItemModel(currentPosition: Int, item: T?): EpoxyModel<*>
+
+  override fun onModelBound(
+    holder: EpoxyViewHolder,
+    boundModel: EpoxyModel<*>,
+    position: Int,
+    previouslyBoundModel: EpoxyModel<*>?
+  ) {
+    // TODO the position may not be a good value if there are too many injected items.
```

Comment:
this seems like it would rarely be a problem, but would be nice to handle. I don't see a great way to handle it well, but here is one general idea

1. Have a way to check whether `boundModel` is of type `T`
2. If so, have a method `modelCache.loadAround(boundModel)`
3. Have the model cache look up the model's position

Speed would be a concern - maybe an approach like this could be async

Reply:
yea not sure what to do here, may get even worse if someone lays out items in reverse etc in `addModels`. I can consider commenting about this in `addModels` but i believe that is distraction as people are unlikely to do something like that. 

we cannot check `boundModel` is type `T` because we don't know what the bound model refers to  (we didn't create it). We can consider maintaining a map from `boundModel` to `position`. Seems like that is the only feasible solution.
Can we do it as a followup ?

Reply:
sure, I think a followup would be great. This is fine for now